### PR TITLE
util/process_wrapper: system_posix.cc: use waitpid correctly

### LINKED ISSUE #708: Compilation errors hidden on Linux/MacOS
Consider the following malformed `main.rs`:

```rust
fn main() {
  println!("ok"
}
```

and the corresponding BUILD file:

```py
load("@rules_rust//rust:rust.bzl", "rust_binary")

rust_binary(
    name = "main",
    srcs = [
        "main.rs",
    ],
)
```

I expect `bazel build //:main` to fail with an error pointing out the missing closing paren. Instead, on Linux/MacOS, all I see is this:

```
$ bazel build //:main 
ERROR: BUILD:3:12: output 'main' was not created
ERROR: BUILD:3:12: Compiling Rust bin main (1 files) failed: not all outputs were created or valid
Target //main:main failed to build
Use --verbose_failures to see the command lines of failed build steps.
FAILED: Build did NOT complete successfully
```

The action correctly fails, but the `rustc` stdout/stderr containing the compilation errors is not printed to the console.

This is a severe usability problem: it means no compiler errors are ever shown to users on Linux/MacOS.

After some debugging, I believe the issue lies in how `system_posix.cc` handles the `rustc` exit status:

https://github.com/bazelbuild/rules_rust/blob/main/util/process_wrapper/system_posix.cc#L177-L182

This code uses `waitpid` incorrectly. The value placed in the `exit_status` variable by `waitpid` is not actually the child's exit status. It is an opaque integer from which the actual exit status can be extracted via the `WEXITSTATUS` macro ([documentation](https://linux.die.net/man/2/waitpid)). As a result, I believe Bazel gets confused about the exit status of the action as a whole, and does not print its stdout/stderr.

When I replace `return exit_status` by `return WEXITSTATUS(exit_status)`, Bazel prints the compilation error to the console as expected.

### PR DESCRIPTION
**Pull request by hlopko**
All the credit goes to https://github.com/bazelbuild/rules_rust/pull/709 and to @Ubehebe.

I'm barely addressing a nit.

Fixes #708.

### CODE REVIEW

**Review by gribozavr — APPROVED** [review]

File: util/process_wrapper/system_posix.cc [hunk 1]
```diff
@@ -174,12 +175,28 @@ int System::Exec(const System::StrType &executable,
...
-  return exit_status;
+  if (WIFEXITED(exit_status)) {
+    return WEXITSTATUS(exit_status);
+  } else if (WIFSIGNALED(exit_status)) {
+    raise(WTERMSIG(exit_status));
+#ifdef WCOREDUMP
+  } else if (WCOREDUMP(exit_status)) {
+    std::cerr << "process wrapper error: child process produced a core dump.\n";
+    return -1;
+#endif
```
Comment (dfreese) [inline util/process_wrapper/system_posix.cc:191]: I don't think this is quite right, given [the man page](https://man7.org/linux/man-pages/man2/wait.2.html):
>WCOREDUMP(wstatus)
              returns true if the child produced a core dump (see
              core(5)).  This macro should be employed only if
              WIFSIGNALED returned true.

Comment (dfreese) [reply util/process_wrapper/system_posix.cc:191]: This seems like it's just additional information, but not strictly necessary.  Can you either incorporated it into the condition on line 185, or remove it?

Comment (hlopko) [reply util/process_wrapper/system_posix.cc]: Good catch, removed.

**Review by dfreese — APPROVED** [review]

**Comment by UebelAndre** [comment]
@hlopko seems like a new warning has appeared as a result of these changes?

From [builds/3211](https://buildkite.com/bazel/rules-rust-rustlang/builds/3211#f7b64823-f793-4ee8-9ff2-607a0ce65a6c)
```
(21:24:44) INFO: From Compiling util/process_wrapper/system_posix.cc:
util/process_wrapper/system_posix.cc: In static member function 'static int process_wrapper::System::Exec(const StrType&, const Arguments&, const EnvironmentBlock&, const StrType&, const StrType&)':
util/process_wrapper/system_posix.cc:195:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
```

**Pull request merged**